File: src/main/java/com/ggardet/patient/spi/mapper/MedicationMapper.java
package com.ggardet.patient.spi.mapper;

import com.ggardet.patient.internal.entity.Medication;
import com.ggardet.patient.spi.dto.MedicationDTO;
import org.mapstruct.Mapper;
import org.mapstruct.MappingConstants;

import java.util.List;

@Mapper(componentModel = MappingConstants.ComponentModel.SPRING)
public interface MedicationMapper {
    MedicationDTO toDto(final Medication medication);
    Medication toEntity(final MedicationDTO medicationDTO);
    List<MedicationDTO> toDtos(final List<Medication> medications);
}

File: src/main/kotlin/com/ggardet/patient/spi/mapper/MedicationMapper.kt
package com.ggardet.patient.spi.mapper

import com.ggardet.patient.internal.entity.Medication
import com.ggardet.patient.spi.dto.MedicationDTO
import org.springframework.stereotype.Component

@Component
class MedicationMapper {
    
    fun toDto(medication: Medication): MedicationDTO {
        return MedicationDTO(
            id = medication.id,
            patientId = medication.patientId!!,
            name = medication.name,
            dosage = medication.dosage,
            frequency = medication.frequency,
            startDate = medication.startDate,
            endDate = medication.endDate,
            prescribingDoctorId = medication.prescribingDoctorId,
            instructions = medication.instructions
        )
    }

    fun toEntity(medicationDTO: MedicationDTO): Medication {
        return Medication(
            id = medicationDTO.id,
            patientId = medicationDTO.patientId,
            name = medicationDTO.name,
            dosage = medicationDTO.dosage,
            frequency = medicationDTO.frequency,
            startDate = medicationDTO.startDate,
            endDate = medicationDTO.endDate,
            prescribingDoctorId = medicationDTO.prescribingDoctorId,
            instructions = medicationDTO.instructions
        )
    }

    fun toDtos(medications: List<Medication>): List<MedicationDTO> {
        return medications.map(::toDto)
    }
}